pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1230,25 +1262,70 @@ func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.
 			// Since this was recovered now, add it to the map, or it might be overwritten
 			taskRunByPipelineTask[pipelineTaskName] = taskrun.Name
 		}
+		if _, ok := childRefByPipelineTask[pipelineTaskName]; minimalEmbedded && !ok {
+			// This taskrun was missing from the status.
+			// Add it without conditions, which are handled in the next loop
+			logger.Infof("Found a TaskRun %s that was missing from the PipelineRun status", taskrun.Name)
+
+			// Since this was recovered now, add it to the map, or it might be overwritten
+			childRefByPipelineTask[pipelineTaskName] = &v1beta1.ChildStatusReference{
+				TypeMeta: runtime.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "TaskRun",
+				},
+				Name:             taskrun.Name,
+				PipelineTaskName: pipelineTaskName,
+			}
+		}
 	}
 	// Then loop by pipelinetask name over all the TaskRuns associated to Conditions
 	for pipelineTaskName, actualConditionTaskRuns := range conditionTaskRuns {
-		taskRunName, ok := taskRunByPipelineTask[pipelineTaskName]
-		if !ok {
+		ok := false
+		taskRunName := ""
+
+		if taskRunName, ok = taskRunByPipelineTask[pipelineTaskName]; !ok && fullEmbedded {
 			// The pipelineTask associated to the conditions was not found in the pipelinerun
 			// status. This means that the conditions were orphaned, and never added to the
 			// status. In this case we need to generate a new TaskRun name, that will be used
 			// to run the TaskRun if the conditions are passed.
-			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pipelineTaskName, pr.Name)
+			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pr.Status.ChildReferences, pipelineTaskName, pr.Name)
 			pr.Status.TaskRuns[taskRunName] = &v1beta1.PipelineRunTaskRunStatus{
 				PipelineTaskName: pipelineTaskName,
 				Status:           nil,
 				ConditionChecks:  nil,
 			}
 		}
+
+		if !ok && minimalEmbedded {
+			if _, ok := childRefByPipelineTask[pipelineTaskName]; !ok {
+				// The pipelineTask associated to the conditions was not found in the pipelinerun
+				// status. This means that the conditions were orphaned, and never added to the
+				// status. In this case we need to generate a new TaskRun name, that will be used
+				// to run the TaskRun if the conditions are passed.
+				taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pr.Status.ChildReferences, pipelineTaskName, pr.Name)
+				childRefByPipelineTask[pipelineTaskName] = &v1beta1.ChildStatusReference{
+					TypeMeta: runtime.TypeMeta{
+						APIVersion: v1beta1.SchemeGroupVersion.String(),
+						Kind:       "TaskRun",
+					},
+					Name:             taskRunName,
+					PipelineTaskName: pipelineTaskName,
+				}
+			}
+		}
+
 		// Build the map of condition checks for the taskrun
 		// If there were no other condition, initialise the map
-		conditionChecks := pr.Status.TaskRuns[taskRunName].ConditionChecks
+		conditionChecks := make(map[string]*v1beta1.PipelineRunConditionCheckStatus)
```

Comment:
if someone has the feature flag set to "both", I think it would be better to treat the childReferences as a source of truth instead of getting condition checks from both statuses. Also, I think it would be worth pulling the condition check logic into a helper function.

Reply:
Good point - I think I may not be consistent on which of `ChildReferences` or `TaskRuns`/`Runs` I use as source of truth if both are populated. I definitely use `ChildReferences` in `cancel.go`, and using the forward-looking one as the source of truth does seem better. I'll rework that, and I'll see if I can tease out the condition check stuff.